Compiler nags with 2.2.0

Congratulations on getting 2.2.0 shipped to CRAN.  Updating my local installation (under Ubuntu 24.10 using the system `libnlopt`) I see

```
checking for pkg-config checking NLopt version... >= 2.7.0
Copying headers found in /usr/include into inst/include
configure: creating ./config.status
config.status: creating src/Makevars
** libs
using C compiler: ‘gcc-14 (Ubuntu 14.2.0-4ubuntu2) 14.2.0’
using C++ compiler: ‘g++-14 (Ubuntu 14.2.0-4ubuntu2) 14.2.0’
ccache g++-14 -I"/usr/share/R/include" -DNDEBUG       -fpic  -O3 -Wall -pipe -pedantic -Wno-ignored-attributes     -c dummy.cpp -o dummy.o
ccache gcc-14 -I"/usr/share/R/include" -DNDEBUG       -fpic  -O3 -Wall -pipe -pedantic -c init_nloptr.c -o init_nloptr.o
ccache gcc-14 -I"/usr/share/R/include" -DNDEBUG       -fpic  -O3 -Wall -pipe -pedantic -c nloptr.c -o nloptr.o
In function ‘compAlg’,
    inlined from ‘bsearch’ at /usr/include/x86_64-linux-gnu/bits/stdlib-bsearch.h:33:23,
    inlined from ‘getVal’ at nloptr.c:122:19:
nloptr.c:118:18: warning: array subscript ‘ALGPAIR {aka const struct algpair}[0]’ is partly outside array bounds of ‘const char[8]’ [-Warray-bounds=]
  118 |   return strcmp(a->key, b->key);
      |                 ~^~~~~
nloptr.c: In function ‘getVal’:
nloptr.c:121:24: note: object ‘key’ of size 8
  121 | int getVal(const char *key) {
      |            ~~~~~~~~~~~~^~~
ccache g++-14 -Wl,-S -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z,relro -o nloptr.so dummy.o init_nloptr.o nloptr.o -llapack -lblas -lgfortran -lm -lquadmath -lnlopt -L/usr/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-nloptr/00new/nloptr/libs
```

Can you take a look?

Also, `Copying headers found in /usr/include into inst/include` seems unusual.  What if I upgrade my Ubuntu system (as I do) and it gets a newer `libnlopt` library and headers?  By creating a copy of the headers we now have potentially re-created a mismatch between header (stale copy in `nloptr`) and library (in system).

--- a/src/nloptr.c
+++ b/src/nloptr.c
@@ -126,7 +126,7 @@ int getVal(char *key) {
 }
 
 // Convert passed string to an nlopt_algorithm item.
-nlopt_algorithm getAlgorithmCode(const char *algorithm_str) {
+nlopt_algorithm getAlgorithmCode(char *algorithm_str) {
 
   nlopt_algorithm algorithm;
 

--- a/src/nloptr.c
+++ b/src/nloptr.c
@@ -118,7 +118,7 @@ static int compAlg(const void *va, const void *vb) {
   return strcmp(a->key, b->key);
 }
 
-int getVal(const char *key) {
+int getVal(char *key) {
   ALGPAIR key_pair[1] = {{key}};
   ALGPAIR *pair = bsearch(key_pair, algtable, sizeof algtable / sizeof algtable[0],
                           sizeof algtable[0], compAlg);

--- a/src/nloptr.c
+++ b/src/nloptr.c
@@ -126,11 +126,11 @@ int getVal(char *key) {
 }
 
 // Convert passed string to an nlopt_algorithm item.
-nlopt_algorithm getAlgorithmCode(char *algorithm_str) {
+nlopt_algorithm getAlgorithmCode(const char *algorithm_str) {
 
   nlopt_algorithm algorithm;
 
-  switch(getVal(algorithm_str)) {
+  switch(getVal((char*)algorithm_str)) {
   case 1:
     algorithm = NLOPT_GD_MLSL;
     break;

--- a/src/nloptr.c
+++ b/src/nloptr.c
@@ -119,7 +119,8 @@ static int compAlg(const void *va, const void *vb) {
 }
 
 int getVal(const char *key) {
-  ALGPAIR *pair = bsearch(&key, algtable, sizeof algtable / sizeof algtable[0],
+  ALGPAIR key_pair[1] = {{key}};
+  ALGPAIR *pair = bsearch(key_pair, algtable, sizeof algtable / sizeof algtable[0],
                           sizeof algtable[0], compAlg);
   return pair ? pair->value : -1;
 }
@@ -196,6 +197,8 @@ nlopt_algorithm getAlgorithmCode(const char *algorithm_str) {
   case 22:
 #ifdef HAVE_NLOPT_LD_LBFGS_NOCEDAL
     algorithm = NLOPT_LD_LBFGS_NOCEDAL;
+#else
+    algorithm = NLOPT_LD_LBFGS;
 #endif
     break;
   case 23:

